prov/rxm/src/rxm_rma.c

```diff
@@ -169,27 +365,36 @@ static ssize_t
 rxm_ep_rma_inject(struct rxm_ep *rxm_ep, const struct fi_msg_rma *msg, uint64_t flags)
 {
 	struct rxm_tx_entry *tx_entry;
-	struct rxm_tx_buf *tx_buf;
-	struct fi_msg_rma msg_rma;
+	struct rxm_rma_buf *rma_buf;
 	struct util_cmap_handle *handle;
 	struct rxm_conn *rxm_conn;
-	struct iovec iov;
-	size_t size;
+	size_t total_size = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
 	ssize_t ret;
 
 	ret = ofi_cmap_get_handle(rxm_ep->util_ep.cmap, msg->addr, &handle);
-	if (OFI_UNLIKELY(ret))
+	if (OFI_UNLIKELY(ret)) {
+		if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+			return ret;
+		fastlock_acquire(&rxm_ep->util_ep.cmap->lock);
+		ret = rxm_ep_format_rma_inject_res(rxm_ep, total_size,
+						   flags, FI_WRITE, msg,
+						   &rma_buf, &tx_entry);
+		if (OFI_LIKELY(!ret)) {
+			rxm_conn = container_of(handle, struct rxm_conn, handle);
+			dlist_insert_tail(&tx_entry->postponed_entry,
+					  &rxm_conn->postponed_tx_list);
+		}
+		fastlock_release(&rxm_ep->util_ep.cmap->lock);
```

Comment:
Can we wrap the acquire lock + insert into postponed_tx_list in a function? Also I think the formatting function called before acquiring the lock.

Reply:
Yep, resources should be allocated outside the section where lock is already acquired
and yes, we can wrap.
Thanks!

Reply:
Is it still not possible to wrap the locking into rxm_ep_postponed_rma function? I think they are needed only around the `dlist_insert()` and once before to check the state and return if needed.

Reply:
Sorry, it is not possible due to the following scenario:
1. main thread (MT) enters to rxm_ep_send_common()
2. (MT) tries to get connection handle (this is unsuccessful -- connection wasn't established yet)
3. (MT) during resources allocation to postpone operation -- connection thread (CT) acquires the lock has got successful connection establishment and then yields the lock.
4. (MT) acquires the lock and inserts the TX entry to the list of deferred operations (this will never be sent)

Reply:
There has to be a check in the pospone_send function where it would check the connection state before inserting the op into the list. That should prevent any deferred op from not being sent. I would prefer this if there is a way to re-use the resources allocated in postpone_send for regular send as well (in case we find the connection has been established).

Nevertheless, I'm fine with the current code and the refactoring can be done later (if possible).

P.S. Github is hiding comments on outdated diff. It took me a while to find this.